File: core/src/net/artux/pda/map/engine/systems/DeadCheckerSystem.java
package net.artux.pda.map.engine.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;

import net.artux.pda.map.DataRepository;
import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.components.HealthComponent;
import net.artux.pda.map.engine.components.InteractiveComponent;
import net.artux.pda.map.engine.components.PositionComponent;
import net.artux.pda.map.engine.components.SpriteComponent;
import net.artux.pda.map.engine.components.StalkerComponent;
import net.artux.pda.map.engine.components.TimeComponent;
import net.artux.pda.map.engine.systems.player.PlayerSystem;
import net.artux.pda.map.ui.UserInterface;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Collections;
import java.util.Random;

import javax.inject.Inject;

@PerGameMap
public class DeadCheckerSystem extends BaseSystem {

    private Group gameZone;
    private Label.LabelStyle labelStyle;
    private boolean deadMessage;
    private DataRepository dataRepository;
    private UserInterface userInterface;

    private ComponentMapper<PositionComponent> pm = ComponentMapper.getFor(PositionComponent.class);
    private ComponentMapper<HealthComponent> hm = ComponentMapper.getFor(HealthComponent.class);

    private AssetManager assetManager;
    private Random random = new Random();

    @Inject
    public DeadCheckerSystem(UserInterface userInterface, DataRepository dataRepository, AssetManager assetManager) {
        super(Family.all(HealthComponent.class, PositionComponent.class).get());
        this.gameZone = userInterface;
        this.assetManager = assetManager;
        this.userInterface = userInterface;
        this.dataRepository = dataRepository;
        labelStyle = userInterface.getLabelStyle();
        labelStyle.fontColor = Color.RED;
    }

    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);

        for (int i = 0; i < getEntities().size(); i++) {
            Entity entity = getEntities().get(i);

            HealthComponent healthComponent = hm.get(entity);
            PositionComponent positionComponent = pm.get(entity);

            if (healthComponent.isDead()) {
                if (random.nextInt(10) < 5) {
                    final Entity deadEntity = new Entity();
                    deadEntity.add(new PositionComponent(positionComponent.getPosition()))
                            .add(new SpriteComponent(assetManager.get("gray.png", Texture.class), 4, 4));

                    if (entity != getPlayer()) {
                        StalkerComponent stalkerComponent = entity.getComponent(StalkerComponent.class);
                        deadEntity.add(new InteractiveComponent("Обыскать: " + stalkerComponent.getName(), 5, new InteractiveComponent.InteractListener() {
                                    @Override
                                    public void interact() {
                                        switch (random.nextInt(4)) {
                                            case 0:
                                                dataRepository.applyActions(Collections.singletonMap("add", Collections.singletonList("85:1")));
                                                break;
                                            case 1:
                                                dataRepository.applyActions(Collections.singletonMap("add", Collections.singletonList("83:2")));
                                                break;
                                            case 2:
                                                dataRepository.applyActions(Collections.singletonMap("add", Collections.singletonList("84:1")));
                                                break;
                                            case 3:
                                                dataRepository.applyActions(Collections.singletonMap("add", Collections.singletonList("85:2")));
                                                break;
                                            case 4:
                                                dataRepository.applyActions(Collections.singletonMap("add", Collections.singletonList("83:1")));
                                                break;
                                        }

                                        getEngine().removeEntity(deadEntity);
                                    }
                                }))
                                .add(new TimeComponent(Instant.now().plus(1, ChronoUnit.MINUTES),
                                        () -> getEngine().removeEntity(entity)))
                                .add(stalkerComponent);
                    } else {
                        getEngine().removeEntity(getPlayer());
                        getEngine().removeSystem(getEngine().getSystem(PlayerSystem.class));
                    }

                    getEngine().addEntity(deadEntity);
                }
                getEngine().removeEntity(entity);
            }
        }
        if (!isPlayerActive()) {
            if (!deadMessage) {
                System.out.println("Dead message from :" + this);
                dataRepository.applyActions(Collections.singletonMap("xp", Collections.singletonList("-5")));
                Group deadMessageGroup = new Group();

                TextButton.TextButtonStyle style = new TextButton.TextButtonStyle();
                style.font = labelStyle.font;
                style.fontColor = Color.RED;

                TextButton textButton = new TextButton("Игра провалена! \n Для продолжения нажмите в любом месте.", style);
                textButton.setSize(Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
                textButton.align(Align.center);
                textButton.getLabel().setAlignment(Align.center);
                deadMessageGroup.addActor(textButton);
                userInterface.addListener(new ClickListener() {
                    @Override
                    public void clicked(InputEvent event, float x, float y) {
                        dataRepository.getPlatformInterface().restart();
                        super.clicked(event, x, y);
                    }
                });
                gameZone.addActor(deadMessageGroup);

                deadMessage = true;
            }
        }
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {

    }

}

File: core/src/net/artux/pda/map/ecs/battle/DeadCheckerSystem.kt
package net.artux.pda.map.ecs.battle

import com.badlogic.ashley.core.ComponentMapper
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.Family
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.physics.box2d.World
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.ui.TextButton.TextButtonStyle
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Align
import net.artux.pda.map.ecs.ai.StalkerComponent
import net.artux.pda.map.ecs.characteristics.HealthComponent
import net.artux.pda.map.ecs.physics.BodyComponent
import net.artux.pda.map.repository.DataRepository
import net.artux.pda.map.ecs.interactive.InteractionSystem
import net.artux.pda.map.ecs.interactive.InteractiveComponent
import net.artux.pda.map.ecs.interactive.TimeComponent
import net.artux.pda.map.ecs.render.SpriteComponent
import net.artux.pda.map.ecs.systems.BaseSystem
import net.artux.pda.map.di.scope.PerGameMap
import net.artux.pda.map.view.window.LootWindow
import net.artux.pda.map.view.root.UserInterface
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import javax.inject.Inject

@PerGameMap
class DeadCheckerSystem @Inject constructor(
    userInterface: UserInterface,
    lootWindow: LootWindow,
    dataRepository: DataRepository,
    assetManager: AssetManager,
    world: World
) : BaseSystem(Family.all(HealthComponent::class.java, BodyComponent::class.java).get()) {

    private val gameZone: Group
    private val lootWindow: LootWindow
    private val labelStyle: LabelStyle
    private var deadMessage = false
    private val dataRepository: DataRepository
    private val userInterface: UserInterface
    private val pm = ComponentMapper.getFor(BodyComponent::class.java)
    private val hm = ComponentMapper.getFor(HealthComponent::class.java)
    private val assetManager: AssetManager
    private val world: World

    init {
        this.assetManager = assetManager
        this.userInterface = userInterface
        this.dataRepository = dataRepository
        this.lootWindow = lootWindow
        this.world = world
        labelStyle = userInterface.labelStyle
        labelStyle.fontColor = Color.RED

        gameZone = userInterface
    }

    override fun update(deltaTime: Float) {
        super.update(deltaTime)
        for (i in 0 until entities.size()) {
            val entity = entities[i]
            val healthComponent = hm[entity]
            val bodyComponent = pm[entity]
            if (healthComponent.isDead()) {

                val deadEntity = Entity()
                deadEntity.add(
                    BodyComponent(
                        bodyComponent.position,
                        world
                    )
                )
                    .add(
                        SpriteComponent(
                            assetManager.get("textures/icons/entity/gray.png", Texture::class.java),
                            4f,
                            4f
                        )
                    )
                if (entity !== player) {
                    val stalkerComponent = entity.getComponent(
                        StalkerComponent::class.java
                    )
                    deadEntity.add(
                        InteractiveComponent(
                            "Обыскать: " + stalkerComponent.name,
                            5
                        ) {
                            lootWindow.updateBot(
                                stalkerComponent.name,
                                stalkerComponent.getAvatar(),
                                stalkerComponent.inventory
                            )
                            userInterface.stack.add(lootWindow)
                            engine.removeEntity(deadEntity)
                        })
                        .add(TimeComponent(
                            Instant.now().plus(1, ChronoUnit.MINUTES)
                        ) { engine.removeEntity(entity) })
                        .add(stalkerComponent)
                } else {
                    engine.getSystem(InteractionSystem::class.java).setProcessing(false)
                    engine.removeEntity(player)
                }
                engine.addEntity(deadEntity)
                engine.removeEntity(entity)
            }
        }
        if (!isPlayerActive) {
            if (!deadMessage) {
                println("Dead message from :$this")
                dataRepository.applyActions(Collections.singletonMap("xp", listOf("-5")), false)
                val style = TextButtonStyle()
                style.font = labelStyle.font
                style.fontColor = Color.RED
                val textButton =
                    TextButton("Игра провалена! \n Для продолжения нажмите в любом месте.", style)
                textButton.setFillParent(true)
                textButton.align(Align.center)
                textButton.label.setAlignment(Align.center)
                userInterface.addListener(object : ClickListener() {
                    override fun clicked(event: InputEvent, x: Float, y: Float) {
                        dataRepository.platformInterface.restart()
                        super.clicked(event, x, y)
                    }
                })
                gameZone.clearChildren()
                gameZone.addActor(textButton)
                deadMessage = true
            }
        }
    }

    override fun processEntity(entity: Entity, deltaTime: Float) {}
}